client/files.go

```diff
@@ -0,0 +1,204 @@
+	GroupID *int
+
+	// Group indicates the name of the owner group for the file in the
+	// destination machine. When used together with MakeDirs, the directories
+	// that might be created will also be owned by this group.
+	Group string
+}
+
+type writeFilesPayload struct {
+	Action string           `json:"action"`
+	Files  []writeFilesItem `json:"files"`
+}
+
+type writeFilesItem struct {
+	Path        string `json:"path"`
+	MakeDirs    bool   `json:"make-dirs"`
+	Permissions string `json:"permissions"`
+	UserID      *int   `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     *int   `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+type fileResult struct {
+	Path  string       `json:"path"`
+	Error *errorResult `json:"error,omitempty"`
+}
+
+type errorResult struct {
+	Message string      `json:"message"`
+	Kind    string      `json:"kind,omitempty"`
+	Value   interface{} `json:"value,omitempty"`
+}
+
+type fileUpload struct {
+	localPath string
+	info      writeFilesItem
+}
+
+type fileUploader struct {
+	pr *io.PipeReader
+	pw *io.PipeWriter
+
+	r *bufio.Reader
+	w *bufio.Writer
+
+	mw    *multipart.Writer
+	files *[]fileUpload
+}
+
+func newFileUploader() (*fileUploader, error) {
+	u := &fileUploader{}
+	u.pr, u.pw = io.Pipe()
+	u.r = bufio.NewReader(u.pr)
+	u.w = bufio.NewWriter(u.pw)
+	u.mw = multipart.NewWriter(u.w)
+
+	return u, nil
+}
+
+func (u *fileUploader) prepareUpload(files []fileUpload) error {
+	// Create metadata form field
+	metaWriter, err := u.mw.CreateFormField("request")
+	if err != nil {
+		return err
+	}
+
+	// Build and encode JSON payload
+	payload := writeFilesPayload{
+		Action: "write",
+		Files:  make([]writeFilesItem, len(files)),
+	}
+	for i, file := range files {
+		payload.Files[i] = file.info
+	}
+
+	encoder := json.NewEncoder(metaWriter)
+	if err := encoder.Encode(&payload); err != nil {
+		return err
+	}
+
+	u.files = &files
+	return nil
+}
+
+func (u *fileUploader) uploadFiles() error {
+	for _, file := range *u.files {
+		w, err := u.mw.CreateFormFile("files", file.info.Path)
+		if err != nil {
+			return err
+		}
+
+		r, err := os.Open(file.localPath)
+		if err != nil {
+			return err
+		}
+		defer r.Close()
+
+		if _, err := io.Copy(w, r); err != nil {
+			return err
+		}
+	}
+	return u.mw.Close()
+}
+
+// PushFile sends a local file to the remote machine.
+func (client *Client) PushFile(opts *PushFileOptions) error {
+	u, err := newFileUploader()
+	if err != nil {
+		return err
+	}
+
+	err = u.prepareUpload([]fileUpload{
+		{
+			localPath: opts.LocalPath,
+			info: writeFilesItem{
+				Path:        opts.RemotePath,
+				MakeDirs:    opts.MakeDirs,
+				Permissions: opts.Permissions,
+				UserID:      opts.UserID,
+				User:        opts.User,
+				GroupID:     opts.GroupID,
+				Group:       opts.Group,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		if err := u.uploadFiles(); err != nil {
+			panic(err)
```

Comment:
Per above, we probably won't need this goroutine, but normally for error handling you'd use a channel to send the err value back to the caller.